File: yoojongho/2024-07-28-combinations.java
/**
 * Leetcode
 * problem: 77
 * link: https://leetcode.com/problems/combinations/description/
 * tag: Backtracking
 */

/**
 * use backtracking
 *
 */

class Solution {
    List<List<Integer>> result;
    int range;
    int size;
    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<>();
        range = n;
        size = k;
        helper(new ArrayList(), 0);
        return result;
    }

    private void helper(List<Integer> temp, int flag){
        flag++;
        if(temp.size() == size){
            result.add(new ArrayList(temp));
            return;
        }
        for(int i = flag; i <= range; i++){
            temp.add(i);
            helper(temp, i);
            temp.remove(temp.size() - 1);
        }
    }
}

File: Soohyeun/2024-07-28-Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        # 1 2 3 4
        # 21 31 41 32 42
        # 312 412 341

        ans = []
        def backTracking(curr, first_num):
            if len(curr) == k:
                ans.append(curr[:])
                return

            for i in range(first_num, n+1):
                curr.append(i)
                backTracking(curr, i + 1)
                curr.pop()

        backTracking([], 1)
        return ans